zebra-state/src/service/read.rs

```diff
@@ -338,62 +382,121 @@ where
+    // - Ok contains the finalized tip height (no need to compensate)
+    // - Err contains the required non-finalized chain overlap
+    let finalized_tip_status = required_min_non_finalized_root..=finalized_tip_range.end().0;
+    let finalized_tip_status = if finalized_tip_status.is_empty() {
+        let finalized_tip_height = *finalized_tip_range.end();
+        Ok(finalized_tip_height)
+    } else {
+        let required_non_finalized_overlap = finalized_tip_status;
+        Err(required_non_finalized_overlap)
+    };
 
     if chain.is_none() {
-        if required_chain_overlap.is_empty() {
-            // The non-finalized chain is empty, and we don't need it.
+        if finalized_tip_status.is_ok() {
+            debug!(
+                ?finalized_tip_status,
+                ?required_min_non_finalized_root,
+                ?finalized_tip_range,
+                ?address_count,
+                "chain address UTXO query: \
+                 finalized chain is consistent, and non-finalized chain is empty",
+            );
+
             return Ok(Default::default());
         } else {
             // We can't compensate for inconsistent database queries,
             // because the non-finalized chain is empty.
-            return Err("unable to get UTXOs: state was committing a block, and non-finalized chain is empty".into());
+            debug!(
+                ?finalized_tip_status,
+                ?required_min_non_finalized_root,
+                ?finalized_tip_range,
+                ?address_count,
+                "chain address UTXO query: \
+                 finalized tip query was inconsistent, but non-finalized chain is empty",
+            );
+
+            return Err("unable to get UTXOs: \
+                        state was committing a block, and non-finalized chain is empty"
+                .into());
         }
     }
 
     let chain = chain.unwrap();
     let chain = chain.as_ref();
 
-    let chain_root = chain.non_finalized_root_height().0;
-    let chain_tip = chain.non_finalized_tip_height().0;
+    let non_finalized_root = chain.non_finalized_root_height();
+    let non_finalized_tip = chain.non_finalized_tip_height();
 
     assert!(
-        chain_root <= required_min_chain_root,
-        "unexpected chain gap: the best chain is updated after its previous root is finalized"
+        non_finalized_root.0 <= required_min_non_finalized_root,
+        "unexpected chain gap: the best chain is updated after its previous root is finalized",
     );
 
-    // If we've already committed this entire chain, ignore its UTXO changes.
-    // This is more likely if the non-finalized state is just getting started.
-    if chain_tip > *required_chain_overlap.end() {
-        if required_chain_overlap.is_empty() {
-            // The non-finalized chain has been committed, and we don't need it.
-            return Ok(Default::default());
```

Comment:
This check always skipped the non-finalized chain.

Instead, we need to:
- skip the non-finalized chain when the finalized tip is consistent and higher
- retry when the finalized tip is inconsistent and higher